Android won't rebuild on config change (Unity 2021)

### Environment

* Unity 2021.2.14f1
* Android (sample project)


### Steps to Reproduce

```pwsh
dotnet msbuild /t:UnityBuildPlayerAndroidIL2CPP /p:Configuration=Release /p:OutDir=other src/Sentry.Unity
dotnet msbuild /t:UnityConfigureSentryOptions src/Sentry.Unity
dotnet msbuild /t:UnityBuildPlayerAndroidIL2CPP /p:Configuration=Release /p:OutDir=other src/Sentry.Unity
```

### Expected Result

A successful build

### Actual Result

```
  /Users/ivan/sentry/sentry-unity/samples/unity-of-bugs/Library/Bee/Android/Prj/IL2CPP/Gradle/unityLibrary/src/main/AndroidManifest.xml:38:5-101 Error:
        Element meta-data#io.sentry.dsn at AndroidManifest.xml:38:5-101 duplicated with element declared at AndroidManifest.xml:31:5-137
  /Users/ivan/sentry/sentry-unity/samples/unity-of-bugs/Library/Bee/Android/Prj/IL2CPP/Gradle/unityLibrary/src/main/AndroidManifest.xml Error:
        Validation failed, exiting
  
  FAILURE: Build failed with an exception.
  
  * What went wrong:
  Execution failed for task ':unityLibrary:processReleaseManifest'.
  > A failure occurred while executing com.android.build.gradle.internal.tasks.Workers$ActionFacade
     > Manifest merger failed with multiple errors, see logs
```

```xml
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android" package="com.unity3d.player" xmlns:tools="http://schemas.android.com/tools">
  <application android:extractNativeLibs="true">
    <activity android:name="com.unity3d.player.UnityPlayerActivity" android:theme="@style/UnityThemeSelector" android:screenOrientation="fullSensor" android:launchMode="singleTask" android:configChanges="mcc|mnc|locale|touchscreen|keyboard|keyboardHidden|navigation|orientation|screenLayout|uiMode|screenSize|smallestScreenSize|fontScale|layoutDirection|density" android:resizeableActivity="false" android:hardwareAccelerated="false">
      <intent-filter>
        <action android:name="android.intent.action.MAIN" />
        <category android:name="android.intent.category.LAUNCHER" />
      </intent-filter>
      <meta-data android:name="unityplayer.UnityActivity" android:value="true" />
      <meta-data android:name="android.notch_support" android:value="true" />
    </activity>
    <meta-data android:name="unity.splash-mode" android:value="0" />
    <meta-data android:name="unity.splash-enable" android:value="True" />
    <meta-data android:name="unity.launch-fullscreen" android:value="True" />
    <meta-data android:name="unity.allow-resizable-window" android:value="False" />
    <meta-data android:name="notch.config" android:value="portrait|landscape" />
    <meta-data android:name="io.sentry.dsn" android:value="https://94677106febe46b88b9b9ae5efd18a00@o447951.ingest.sentry.io/5439417" />
    <meta-data android:name="io.sentry.debug" android:value="true" />
    <meta-data android:name="io.sentry.release" android:value="unity-of-bugs@0.1" />
    <meta-data android:name="io.sentry.environment" android:value="production" />
    <meta-data android:name="io.sentry.debug.level" android:value="debug" />
    <meta-data android:name="io.sentry.auto-session-tracking.enable" android:value="False" />
    <meta-data android:name="io.sentry.ndk.scope-sync.enable" android:value="True" />
    <meta-data android:name="io.sentry.dsn" android:value="http://publickey@localhost:8000/12345" />
    <meta-data android:name="io.sentry.debug" android:value="true" />
    <meta-data android:name="io.sentry.release" android:value="unity-of-bugs@0.1" />
    <meta-data android:name="io.sentry.environment" android:value="production" />
    <meta-data android:name="io.sentry.debug.level" android:value="debug" />
    <meta-data android:name="io.sentry.auto-session-tracking.enable" android:value="False" />
    <meta-data android:name="io.sentry.ndk.scope-sync.enable" android:value="True" />
  </application>
  <uses-feature android:glEsVersion="0x00020000" />
  <uses-permission android:name="android.permission.INTERNET" />
  <uses-feature android:name="android.hardware.touchscreen" android:required="false" />
  <uses-feature android:name="android.hardware.touchscreen.multitouch" android:required="false" />
  <uses-feature android:name="android.hardware.touchscreen.multitouch.distinct" android:required="false" />
</manifest>
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,7 @@
 ### Fixes
 
 - Whitespaces no longer cause issues when uploading symbols for Windows native  ([#655](https://github.com/getsentry/sentry-unity/pull/655))
+- AndroidManifest update removes previous `io.sentry` entries ([#652](https://github.com/getsentry/sentry-unity/pull/652))
 
 ## 0.13.0
 

--- a/samples/unity-of-bugs/Assets/Editor/AndroidManifestUpdater.cs
+++ b/samples/unity-of-bugs/Assets/Editor/AndroidManifestUpdater.cs
@@ -16,15 +16,23 @@ public class AndroidManifestUpdater : IPostGenerateGradleAndroidProject
         }
 
         Debug.Log($"AndroidManifestUpdater: Updating {manifestName}");
+        var replacement = "<application android:usesCleartextTraffic=\"true\"";
         var text = File.ReadAllText(manifestPath);
-        var newText = text.Replace("<application", "<application android:usesCleartextTraffic=\"true\"");
-        if (text.Equals(newText))
+        if (text.Contains(replacement))
         {
-            Debug.LogError($"AndroidManifestUpdater: Failed find the <application> tag");
+            Debug.Log($"AndroidManifestUpdater: Already contains the configuration, nothing to do.");
         }
         else
         {
-            File.WriteAllText(manifestPath, newText);
+            var newText = text.Replace("<application", replacement);
+            if (text.Equals(newText))
+            {
+                Debug.LogError($"AndroidManifestUpdater: Failed to update the <application> tag");
+            }
+            else
+            {
+                File.WriteAllText(manifestPath, newText);
+            }
         }
     }
 }

--- a/src/Sentry.Unity.Editor/Android/AndroidManifestConfiguration.cs
+++ b/src/Sentry.Unity.Editor/Android/AndroidManifestConfiguration.cs
@@ -1,4 +1,5 @@
 using System;
+using System.Collections.Generic;
 using System.IO;
 using System.Text;
 using System.Xml;
@@ -84,7 +85,8 @@ namespace Sentry.Unity.Editor.Android
                 disableAutoInit = true;
             }
 
-            var androidManifest = new AndroidManifest(manifestPath);
+            var androidManifest = new AndroidManifest(manifestPath, options?.DiagnosticLogger);
+            androidManifest.RemovePreviousConfigurations();
 
             if (disableAutoInit)
             {
@@ -201,6 +203,7 @@ namespace Sentry.Unity.Editor.Android
     {
         private readonly string _path;
         protected const string AndroidXmlNamespace = "http://schemas.android.com/apk/res/android";
+        protected const string AndroidNsPrefix = "android";
 
         protected AndroidXmlDocument(string path)
         {
@@ -227,35 +230,64 @@ namespace Sentry.Unity.Editor.Android
 
     internal class AndroidManifest : AndroidXmlDocument
     {
+        private const string _prefix = "io.sentry";
         private readonly XmlElement _applicationElement;
+        private readonly IDiagnosticLogger? _logger;
 
-        public AndroidManifest(string path) : base(path) =>
-            _applicationElement = (XmlElement)SelectSingleNode("/manifest/application");
+        public AndroidManifest(string path, IDiagnosticLogger? logger) : base(path)
+        {
+            _applicationElement = (XmlElement) SelectSingleNode("/manifest/application");
+            _logger = logger;
+        }
+
+        internal void RemovePreviousConfigurations()
+        {
+            var nodesToRemove = new List<XmlNode>();
+            foreach (XmlNode node in _applicationElement.ChildNodes)
+            {
+                if (node.Name.Equals("meta-data"))
+                {
+                    foreach (XmlAttribute attr in node.Attributes)
+                    {
+                        if (attr.Prefix.Equals(AndroidNsPrefix) && attr.LocalName.Equals("name") && attr.Value.StartsWith(_prefix))
+                        {
+                            _logger?.LogDebug("Removing AndroidManifest meta-data '{0}'", attr.Value);
+                            nodesToRemove.Add(node);
+                            break;
+                        }
+                    }
+                }
+            }
+            foreach (XmlNode node in nodesToRemove)
+            {
+                _ = node.ParentNode.RemoveChild(node);
+            }
+        }
 
         // Without this we get:
         // Unable to get provider io.sentry.android.core.SentryInitProvider: java.lang.IllegalArgumentException: DSN is required. Use empty string to disable SDK.
         internal void DisableSentryAndSave()
         {
-            SetMetaData("io.sentry.auto-init", "false");
+            SetMetaData($"{_prefix}.auto-init", "false");
             _ = Save();
         }
 
-        internal void SetDsn(string dsn) => SetMetaData("io.sentry.dsn", dsn);
-        internal void SetSampleRate(float sampleRate) => SetMetaData("io.sentry.sample-rate", sampleRate.ToString());
-        internal void SetRelease(string release) => SetMetaData("io.sentry.release", release);
-        internal void SetEnvironment(string environment) => SetMetaData("io.sentry.environment", environment);
+        internal void SetDsn(string dsn) => SetMetaData($"{_prefix}.dsn", dsn);
+        internal void SetSampleRate(float sampleRate) => SetMetaData($"{_prefix}.sample-rate", sampleRate.ToString());
+        internal void SetRelease(string release) => SetMetaData($"{_prefix}.release", release);
+        internal void SetEnvironment(string environment) => SetMetaData($"{_prefix}.environment", environment);
 
         internal void SetAutoSessionTracking(bool enableAutoSessionTracking)
-            => SetMetaData("io.sentry.auto-session-tracking.enable", enableAutoSessionTracking.ToString());
+            => SetMetaData($"{_prefix}.auto-session-tracking.enable", enableAutoSessionTracking.ToString());
 
         internal void SetNdkScopeSync(bool enableNdkScopeSync)
-            => SetMetaData("io.sentry.ndk.scope-sync.enable", enableNdkScopeSync.ToString());
+            => SetMetaData($"{_prefix}.ndk.scope-sync.enable", enableNdkScopeSync.ToString());
 
-        internal void SetDebug(bool debug) => SetMetaData("io.sentry.debug", debug ? "true" : "false");
+        internal void SetDebug(bool debug) => SetMetaData($"{_prefix}.debug", debug ? "true" : "false");
 
         // https://github.com/getsentry/sentry-java/blob/db4dfc92f202b1cefc48d019fdabe24d487db923/sentry/src/main/java/io/sentry/SentryLevel.java#L4-L9
         internal void SetLevel(SentryLevel level) =>
-            SetMetaData("io.sentry.debug.level", level switch
+            SetMetaData($"{_prefix}.debug.level", level switch
             {
                 SentryLevel.Debug => "debug",
                 SentryLevel.Error => "error",
@@ -275,7 +307,7 @@ namespace Sentry.Unity.Editor.Android
 
         private XmlAttribute CreateAndroidAttribute(string key, string value)
         {
-            var attr = CreateAttribute("android", key, AndroidXmlNamespace);
+            var attr = CreateAttribute(AndroidNsPrefix, key, AndroidXmlNamespace);
             attr.Value = value;
             return attr;
         }

--- a/src/Sentry.Unity.Editor/Android/AndroidManifestConfiguration.cs
+++ b/src/Sentry.Unity.Editor/Android/AndroidManifestConfiguration.cs
@@ -236,7 +236,7 @@ namespace Sentry.Unity.Editor.Android
 
         public AndroidManifest(string path, IDiagnosticLogger? logger) : base(path)
         {
-            _applicationElement = (XmlElement) SelectSingleNode("/manifest/application");
+            _applicationElement = (XmlElement)SelectSingleNode("/manifest/application");
             _logger = logger;
         }
 

--- a/test/Sentry.Unity.Editor.Tests/Android/AndroidManifestConfigurationTests.cs
+++ b/test/Sentry.Unity.Editor.Tests/Android/AndroidManifestConfigurationTests.cs
@@ -261,6 +261,41 @@ namespace Sentry.Unity.Editor.Tests.Android
         }
 
         [Test]
+        public void ModifyManifest_RepeatedRunProducesSameResult()
+        {
+            var sut = _fixture.GetSut();
+            var manifest1 = WithAndroidManifest(basePath => sut.ModifyManifest(basePath));
+            var manifest2 = WithAndroidManifest((basePath) =>
+            {
+                sut.ModifyManifest(basePath);
+                sut.ModifyManifest(basePath);
+            });
+
+            Debug.Log(manifest2);
+            Assert.True(manifest1.Contains("sentry.dsn"));
+            Assert.AreEqual(manifest1, manifest2);
+        }
+
+        [Test]
+        public void ModifyManifest_RepeatedRunRemovesConfigs()
+        {
+            var fixture2 = new Fixture();
+            fixture2.SentryUnityOptions = null;
+            var manifest1 = WithAndroidManifest(basePath => _fixture.GetSut().ModifyManifest(basePath));
+            var manifest2 = WithAndroidManifest((basePath) =>
+            {
+                _fixture.GetSut().ModifyManifest(basePath);
+                fixture2.GetSut().ModifyManifest(basePath);
+            });
+
+            Debug.Log($"Manifest 1 (before):\n{manifest1}");
+            Debug.Log($"Manifest 2 (after):\n{manifest2}");
+            Assert.True(manifest1.Contains("sentry.dsn"));
+            Assert.False(manifest2.Contains("sentry.dsn"));
+        }
+
+
+        [Test]
         public void SetupSymbolsUpload_ScriptingBackendNotIL2CPP_LogsAndReturns()
         {
             _fixture.ScriptingImplementation = ScriptingImplementation.Mono2x;

